File: mailblastr-java/src/test/java/com/mailblastr/tests/RecoveryTest.java
package com.mailblastr.tests;

import com.mailblastr.Mailblastr;
import com.mailblastr.MailblastrException;
import com.mailblastr.http.DefaultHttpTransport;
import com.mailblastr.http.HttpResult;
import com.mailblastr.json.Json;
import com.mailblastr.requests.ForwardEmailRequest;
import com.mailblastr.requests.ReplyEmailRequest;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public final class RecoveryTest {
    public static void run() throws Exception {
        Check.suite("Recovery");
        Path root = Path.of("").toAbsolutePath();
        while (!Files.exists(root.resolve("scripts/http-recovery-corpus.json"))) root = root.getParent();
        List<?> cases = (List<?>) Json.parse(Files.readString(root.resolve("scripts/http-recovery-corpus.json")));
        for (Object item : cases) {
            Map<?, ?> c = (Map<?, ?>) item;
            int status = ((Number) c.get("status")).intValue();
            int attempts = ((Number) c.get("attempts")).intValue();
            String key = (String) c.get("key");
            List<String> keys = Collections.synchronizedList(new ArrayList<>());
            List<String> bodies = Collections.synchronizedList(new ArrayList<>());
            HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
            server.createContext("/", exchange -> {
                keys.add(exchange.getRequestHeaders().getFirst("Idempotency-Key"));
                bodies.add(new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8));
                byte[] response = (keys.size() == 1 ? Json.write(c.get("body")) : "{\"id\":\"em_retry\"}").getBytes(StandardCharsets.UTF_8);
                exchange.getResponseHeaders().set("Retry-After", "0");
                exchange.sendResponseHeaders(keys.size() == 1 ? status : 200, response.length);
                exchange.getResponseBody().write(response);
                exchange.close();
            });
            server.start();
            try {
                String payload = "{\"text\":\"a  b\\n\\n c\"}";
                HttpResult result = new DefaultHttpTransport(Duration.ofSeconds(2), 1).execute(
                    (String) c.get("method"), "http://127.0.0.1:" + server.getAddress().getPort() + c.get("path"),
                    key == null ? Collections.emptyMap() : Map.of("Idempotency-Key", key), payload);
                Check.eq(c.get("name") + " attempts", attempts, keys.size());
                Check.eq(c.get("name") + " status", attempts == 1 ? status : 200, result.statusCode());
                Check.isTrue(c.get("name") + " same payload", bodies.stream().allMatch(payload::equals));
                if (key != null) Check.isTrue(c.get("name") + " same key", keys.stream().allMatch(key::equals));
            } finally { server.stop(0); }
        }

        StubTransport t = new StubTransport();
        Mailblastr mb = new Mailblastr("mb_test", "https://api.test", t);
        t.respond(200, "{\"id\":\"em_one\"}");
        mb.emails().receiving().reply("rcv_one", ReplyEmailRequest.builder().text("a  b\n\n c").build(), "reply-1");
        Check.eq("reply key", "reply-1", t.lastHeaders.get("Idempotency-Key"));
        Check.eq("reply whitespace", "a  b\n\n c", ((Map<?, ?>) Json.parse(t.lastBody)).get("text"));
        mb.emails().receiving().forward("rcv_one", ForwardEmailRequest.builder().build(), "forward-1");
        Check.eq("forward key", "forward-1", t.lastHeaders.get("Idempotency-Key"));
        t.respond(200, "{\"custom_host\":null,\"status\":\"shared\",\"checked_at\":\"2026-09-10T00:00:00Z\"}");
        Check.eq("tracking health", "shared", mb.domains().trackingHealth("domain one").getString("status"));
        Check.eq("tracking health path", "https://api.test/domains/domain%20one/tracking-health", t.lastUrl);
        t.respond(422, "{\"statusCode\":503,\"name\":\"validation_error\",\"id\":\"em_one\",\"reserved\":[{\"id\":\"em_one\"}],\"unsent_count\":0}");
        try { mb.emails().get("em_one"); Check.isTrue("expected API error", false); }
        catch (MailblastrException e) {
            Check.eq("actual HTTP status", 422, e.getStatusCode());
            Check.eq("original ID", "em_one", e.getId());
            Check.eq("reserved ID", "em_one", e.getReserved().get(0).get("id"));
            Check.eq("never attempted count", 0, e.getUnsentCount());
        }
    }
}

File: mailblastr-php/tests/cases/RecoveryTest.php
<?php
declare(strict_types=1);

class RecoveryTransport extends \Mailblastr\Transport\CurlTransport
{
    public int $calls = 0;
    public array $requests = [];
    public function __construct(private array $case) { parent::__construct(30, 2); }
    protected function perform(string $method, string $url, array $headers, ?string $body): array
    {
        $this->requests[] = compact('method','url','headers','body');
        $this->calls++;
        return ['status'=>$this->calls===1 ? $this->case['status'] : 200,
            'body'=>json_encode($this->calls===1 ? $this->case['body'] : ['id'=>'em_retry']), 'retryAfter'=>'0'];
    }
    protected function sleepSeconds(float $seconds): void { }
}

$localCorpus = file_get_contents(__DIR__ . '/../fixtures/http-recovery-corpus.json');
$sharedCorpusPath = __DIR__ . '/../../../scripts/http-recovery-corpus.json';
if (is_file($sharedCorpusPath)) {
    check_same('recovery: package corpus matches source', file_get_contents($sharedCorpusPath), $localCorpus);
}
$cases = json_decode($localCorpus, true);
foreach ($cases as $case) {
    $transport = new RecoveryTransport($case);
    $headers = isset($case['key']) ? ['Idempotency-Key: '.$case['key']] : [];
    $result = $transport->request($case['method'], 'http://127.0.0.1'.$case['path'], $headers, '{}');
    check_same('recovery: '.$case['name'].' attempts', $case['attempts'], $transport->calls);
    check_same('recovery: '.$case['name'].' status', $case['attempts']===1 ? $case['status'] : 200, $result['status']);
    foreach ($transport->requests as $request) check_same('recovery: stable headers '.$case['name'], $headers, $request['headers']);
}

[$mb, $t] = make_client();
$t->queue(200, ['object'=>'email','id'=>'em_one']);
$mb->emails->receiving->reply('rcv_one', ['text'=>"a  b\n\n c"], ['idempotencyKey'=>'reply-1']);
check('recovery: reply key', in_array('Idempotency-Key: reply-1', $t->last()['headers'], true));
check_same('recovery: whitespace unchanged', "a  b\n\n c", $t->lastJson()['text']);
$t->queue(200, ['object'=>'email','id'=>'em_one']);
$mb->emails->receiving->forward('rcv_one', [], ['idempotencyKey'=>'forward-1']);
check('recovery: forward key', in_array('Idempotency-Key: forward-1', $t->last()['headers'], true));
$t->queue(200, ['custom_host'=>null,'status'=>'shared','checked_at'=>'2026-09-10T00:00:00Z']);
check_same('recovery: tracking status', 'shared', $mb->domains->trackingHealth('domain one')['status']);
check_same('recovery: tracking path', '/domains/domain%20one/tracking-health', $t->lastPath());
$error = \Mailblastr\Exceptions\MailblastrException::fromResponse(422, ['statusCode'=>503,'id'=>'em_one','reserved'=>[['id'=>'em_one']],'unsent_count'=>0]);
check_same('recovery: HTTP status wins', 422, $error->getStatusCode());
check_same('recovery: original ID', 'em_one', $error->getId());
check_same('recovery: reserved prefix', [['id'=>'em_one']], $error->getReserved());
check_same('recovery: zero unattempted preserved', 0, $error->getUnsentCount());
